Update language-ruby.cson

mamenama opened the pull request:
- Changed 'initialize' prefix to unique 'defi' and added multiple cursor for assignment of instance variable with same name as argument.
- Modified 'heredoc with indented terminator' to 'heredoc with optional multiple inputs and indented terminators'
- Fix for issue where erb blocks would not expand after an HTML tag 
- Fix for formatting of def method *args
- Fix for `||` pipes within block arguments
- Fix for multiple method arguments
- Added grammar styling for symbol hashkeys within method parameters
- Added grammar styling for stabby lambda (->) operator

mamenama commented:
Addressing #117 stabby Lambda (->) operator grammar

MaximSokolov commented on grammars/ruby.cson:
```diff
@@ -1940,6 +1952,10 @@
   }
   {
+    'match': '->'
+    'name': 'keyword.operator.lambda.ruby'
```
JS, CoffeeScript tokenize `->` as `storage.type.function.arrow`

mamenama replied:
Yes - my understanding is that in CoffeeScript the arrows stand for the `function` keyword with (`=>`) and without (`->`) binding (w/ JS also having `=>` but no `->`). In Ruby the `->` is an operator that serves as a shortcut to the Kernel#lambda method, so it should be tokenized differently than in CS. That said, there may be a better match than 'keyword.operator.lambda.ruby' if people have other ideas.

MaximSokolov replied:
>  In Ruby the `->` is an operator that serves as a shortcut to the Kernel#lambda method

It should be tokenzied same as `lambda` keyword then (`support.function.kernel.arrow`)
Don't known why it is tokenized as `support.function`. As far as I understand, `lambda` and `->` in Ruby "define" methods/functions i.e. related to "storage" (`lambda` in python is `storage.type.function`)

mamenama replied:
Yes - I agree it should be the same as `lambda`. Perhaps `lambda` is miscategorized, since it is part of core and not support? How about `storage.type.function.arrow` and change `lambda` to storage.type.function?

MaximSokolov replied:
> Perhaps `lambda` is miscategorized, since it is part of core and not support?

`support.function` is for functions provided by the framework/library (e.g. `puts`)
See [TextMate naming conventions for details](http://manual.macromates.com/en/language_grammars#naming_conventions)

> How about `storage.type.function.arrow` (as you suggested) and change `lambda` to `storage.type.function`

`storage.type` is correct scopes for them. However I almost forget that changing root scopes will change syntax highlining:

<!-- -->
> ![atom_dark](https://cloud.githubusercontent.com/assets/2943616/11005191/81bc6604-84cd-11e5-8628-8a56c97d198f.png)
_Atom Dark_

<!-- -->
> ![one_dark](https://cloud.githubusercontent.com/assets/2943616/11005228/b62b1534-84cd-11e5-9096-264e76dcf7d9.png)
_One Dark_

<!-- -->
> ![monokai](https://cloud.githubusercontent.com/assets/2943616/11005242/c70231c6-84cd-11e5-91b4-0f7060e122b3.png)
_Monokai_

<!-- -->
> ![seti](https://cloud.githubusercontent.com/assets/2943616/11005246/cf1894c2-84cd-11e5-82c3-b017419f2fb9.png)
_Seti_

It seems `keyword.other` for `lambda`, `->` look most closer for existing Ruby syntax highlighting

mamenama replied:
OK - I've added a4f42cf455f105e6fbe4cc2d2bcaf428dccc28a4 to change `->` to match `lambda`. After looking at it further, it seems that all the other Kernel#methods are also in `support.function.kernel.ruby` so it makes sense to keep `lambda` in that category as well.

ioquatix commented:
+1 looks good.

50Wliu commented:
Hey @mamenama, in the future can you please create separate PRs for separate issues?  It makes reviewing a lot easier (and so that the entire PR doesn't have to wait on one small change for one of the issues).  Thanks!

mamenama commented:
Hi @50Wliu - will do. Rookie first-timer mistake - I hadn't realized additional pushes to `fork` on my master would automatically be incorporated into the same PR until the third or fourth one, and then once I'd started down the dark path, well, I got carried away.  I will make any new issues into a separate PR (aside from changes to the above issues).

50Wliu commented on spec/ruby-spec.coffee:
```diff
@@ -532,3 +532,68 @@ describe "Ruby grammar", ->
+
+  it "tokenizes a method with *args properly", ->
+    {tokens} = grammar.tokenizeLine('def method(*args)')
+    expect(tokens[0]).toEqual value: 'def', scopes: [ 'source.ruby', 'meta.function.method.with-arguments.ruby', 'keyword.control.def.ruby' ]
```
:art: Some whitespace inconsistencies in these new specs: I would stick with `['source.ruby', ... 'last.scope.ruby']` since that's what all the other specs use.

50Wliu commented on spec/ruby-spec.coffee:
```diff
@@ -532,3 +532,68 @@ describe "Ruby grammar", ->
+    expect(tokens[9]).toEqual value: '2', scopes: ['source.ruby', 'meta.function.method.with-arguments.ruby', 'constant.numeric.ruby']
+    expect(tokens[12]).toEqual value: 'green', scopes: ['source.ruby', 'meta.function.method.with-arguments.ruby', 'constant.other.symbol.hashkey.parameter.function.ruby']
+    expect(tokens[15]).toEqual value: 'rand', scopes: ['source.ruby', 'meta.function.method.with-arguments.ruby', 'support.function.kernel.ruby']
+    expect(tokens[17]).toEqual value: '3', scopes: ['source.ruby', 'meta.function.method.with-arguments.ruby', 'constant.numeric.ruby']
```
Can you check all the parentheses please?

mamenama commented:
OK @50Wliu - added commits 101e2ba and dc47018 addressing those two issues. Thanks!

50Wliu commented:
Looks good to me.

I'd like someone else to review it as well though because I haven't been keeping up with all the language-ruby issues; /cc @atom/feedback

ioquatix commented:
+1 looks good.

50Wliu commented:
Going to merge this and see how it goes.

@mamenama, thanks for putting all the effort into this.

envygeeks commented:
:heart:

ioquatix commented:
Thanks everyone this is awesome!!

ioquatix commented:
I'm not sure if I'm doing something wrong.. but using 0.64.0 and I get the following:

![image](https://cloud.githubusercontent.com/assets/30030/11326910/da02af7a-91da-11e5-8893-eea5858505b4.png)

Here is the source code:

![image](https://cloud.githubusercontent.com/assets/30030/11326911/e22cff70-91da-11e5-97b2-c9a145a68c86.png)

ioquatix commented:
OH... wait, something is wrong with atom. It says 0.64.0 installed but it appears to be using 0.60.0.. odd.

ioquatix commented:
Hmm, I can't seem to update. In "Install" the latest version is shown:

![image](https://cloud.githubusercontent.com/assets/30030/11326917/2e07202e-91db-11e5-99cb-f365ff5db6ba.png)

But in updates, it's not listed:

![image](https://cloud.githubusercontent.com/assets/30030/11326918/397b1e56-91db-11e5-9c14-694c69760265.png)

However the actual installed version is 0.60.0:

![image](https://cloud.githubusercontent.com/assets/30030/11326920/4ca1b134-91db-11e5-8534-9f0c5b38b622.png)

This seems a bit odd.. any ideas anyone?

50Wliu commented:
Atom pins its bundled packages to specific versions in each release (in this case, Atom 1.2.x is using language-ruby v0.60.0).  To override that execute `apm install language-ruby` from the command line (though this is highly unrecommended).

ioquatix commented:
Okay figured out I need to run `apm install language-ruby@0.64.0` and now it is working, including the ligatures. Thanks.

ioquatix commented:
Hey, so after installing this, one of my own packages isn't rendering well: https://github.com/ioquatix/language-trenni/blob/master/grammars/trenni.cson

ioquatix commented:
Not sure if it's something I'm doing wrong, but perhaps this PR is responsible and might break other things?

![image](https://cloud.githubusercontent.com/assets/30030/11328309/c3c9a4e0-91ec-11e5-8794-5a183ec9e3e0.png)

Notice the `<?r` and `?>` tags are not highlighted correctly and the `#{...}` blocks are now showing up as comments.

50Wliu commented:
The first `#{` block actually isn't.  I suspect that something is going wrong in the `<?r f.select` line.

/cc @mamenama

ioquatix commented:
Okay here are some interesting observations. Both okay:

![image](https://cloud.githubusercontent.com/assets/30030/11328835/f544165e-91f6-11e5-820f-75d1fc6cfa9a.png)

Something wrong:

![image](https://cloud.githubusercontent.com/assets/30030/11328836/fe475860-91f6-11e5-9345-fb140a4c6d4f.png)

Minimal example:

![image](https://cloud.githubusercontent.com/assets/30030/11328841/0b132ca4-91f7-11e5-8694-471faa804ccf.png)

mamenama commented:
OK - I added #123 which looks like it fixes the issue while still retaining the new functionality. The problem stemmed from adding support for `do | a = b||c |` from #93. The original end regex ate the surrounding area and caused havoc. I have replaced it with one that is more innocuous while still retaining the functionality.

@ioquatix - I tested it a bit with your language-trenni package and it looks OK to me now but let me know if it isn't.

ioquatix commented:
@mamenama how do I test this? Check out the code into the package directory?

50Wliu commented:
@ioquatix `git clone` #123 (Github Desktop should allow you to do this easily) and then `apm link` it.
